parsers/mcmaster.js

```diff
@@ -1,283 +1,146 @@
+    var _locationRegex = /(\s[\-\w ,]+)/;
+    var _profRegex = /((\b[A-Za-z]+\s)(?:[A-Za-z\-.]+\s?)+|(\b[A-Za-z]+\s))/;
+    var _locationProfPair = new RegExp(_locationRegex.source + _wsRegex.source + _profRegex.source);
+    var _semesterRegex = /((?:\d{2}\/\d{2}\/\d{4})|(?:\d{4}\/\d{2}\/\d{2})|(?:\d{4}-\d{2}-\d{2}))/;
+    var _semesterPairRegex = new RegExp(_semesterRegex.source +  _wsRegex.source + "-" +  _wsRegex.source +  _semesterRegex.source);
+    var _sectionTypeRegex = /(\d+)\s+([A-Z]{1}\d+)\s+(?=Lecture|Tutorial|Laboratory)([A-Za-z]{3,}\b)/;
+    var _dayRegex = /(Mo)|(Tu)|(We)|(Th)|(Fr)|(Sa)|(Su)/g;
+    var _weekdays = {"Mo":"Monday", "Tu":"Tuesday", "We":"Wednesday", "Th":"Thursday", "Fr":"Friday", "Sa":"Saturday", "Su":"Sunday"};
+    var timetable = [];
+    // Breaks all the courses up. Looks at all the detail after a course name
+    var courses = text.split(_courseTitleLookAhead);
+
+    for (var i = 0; i < courses.length; i++) {
+        var courseText = courses[i];
+        var course = _courseTitleRegex.exec(courseText);
+        // Breaks all classes up, looks at all the details after the Course Code or time in a course
+        var classes = courseText.split(_timeDayPairOrCourseCodeLookAhead);
+
+        if (!course) continue;
+
+        var currentSection = _sectionTypeRegex.exec(courseText);
+
+        for (var j = 0; j < classes.length; j++) {
+            var locationProfPair = _locationProfPair.exec(classes[j]);
+            var time = _timeDayPairRegex.exec(classes[j]);
+            var semester = _semesterPairRegex.exec(classes[j]);
+            var section = _sectionTypeRegex.exec(classes[j]);
+            // Check is the current classes segment is a section if it is update the current Section
+            currentSection = section ? section : currentSection;
+
+            if (!locationProfPair || !time || !semester) continue;
+
+            semester = fixDateFormat(semester[1], semester[2]);
+            var where = locationProfPair[1];
+            var professor = locationProfPair[2];
+            // Changes McMasters MoTuWe format to full weekdays
+            var days = time[1].replace(_dayRegex, function fullWeekday(x){return _weekdays[x] + ",";});
+            var daysArray = days.substring(0, days.length-1).split(",");
+
+            for (var k = 0; k < daysArray.length ; k++) {
+                var startTimeSplit = convertTo24Hour(time[2]).split(":");
+                var startTimeObj = {'hour': parseInt(startTimeSplit[0], 10), 'minute': parseInt(startTimeSplit[1], 10)};
+                var endTimeSplit = convertTo24Hour(time[3]).split(":");
+                var endTimeObj = {'hour': parseInt(endTimeSplit[0], 10), 'minute': parseInt(endTimeSplit[1], 10)};
+                var sem_start = new Date(semester[0]);
+                var sem_end = new Date(semester[1]);
+                var semesterStartObj = { year: sem_start.getFullYear(), month: sem_start.getMonth(), day: sem_start.getDate()};
```

Comment:
put these directly into the `timetable.push` call similar to what I did, no point in creating temporary objects